facilitator/src/transport/gcs.rs

```diff
@@ -0,0 +1,601 @@
+    mem,
+};
+
+const STORAGE_API_BASE_URL: &str = "https://storage.googleapis.com";
+const DEFAULT_OAUTH_TOKEN_URL: &str =
+    "http://metadata.google.internal:80/computeMetadata/v1/instance/service-accounts/default/token";
+
+// A wrapper around an Oauth token and its expiration date.
+#[derive(Debug)]
+struct OauthToken {
+    token: String,
+    expiration: DateTime<Utc>,
+}
+
+impl OauthToken {
+    // Returns true if the token is not yet expired.
+    fn expired(&self) -> bool {
+        Utc::now() < self.expiration
+    }
+}
+
+// Represents the response from a GET request to the GKE metadata service's
+// service account token endpoint. Structure is derived from empirical
+// observation of the JSON scraped from inside a GKE job.
+#[derive(Debug, Deserialize, PartialEq)]
+struct MetadataServiceTokenResponse {
+    access_token: String,
+    expires_in: i64,
+    token_type: String,
+}
+
+// Represents the response from a POST request to the GCP IAM service's
+// generateAccessToken endpoint.
+// https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken
+#[derive(Debug, Deserialize, PartialEq)]
+#[serde(rename_all = "camelCase")]
+struct GenerateAccessTokenResponse {
+    access_token: String,
+    expire_time: DateTime<Utc>,
+}
+
+// OauthTokenProvider manages a default service account Oauth token (i.e. the
+// one for a GCP service account mapped to a Kubernetes service account) and an
+// Oauth token used to impersonate another service account.
+struct OauthTokenProvider {
+    service_account_to_impersonate: Option<String>,
+    default_service_account_oauth_token: Option<OauthToken>,
+    impersonated_service_account_oauth_token: Option<OauthToken>,
+}
+
+impl OauthTokenProvider {
+    // Creates a token provider which can impersonate the specified service
+    // account.
+    fn new(service_account_to_impersonate: Option<String>) -> OauthTokenProvider {
+        OauthTokenProvider {
+            service_account_to_impersonate,
+            default_service_account_oauth_token: None,
+            impersonated_service_account_oauth_token: None,
+        }
+    }
+
+    fn ensure_storage_access_oauth_token(&mut self) -> Result<String> {
+        match self.service_account_to_impersonate {
+            Some(_) => self.ensure_impersonated_service_account_oauth_token(),
+            None => self.ensure_default_service_account_oauth_token(),
+        }
+    }
+
+    // Obtains an Oauth token for the default service account, if one is not
+    // already held or if it has expired. Otherwise returns a copy of the token.
+    // The returned value is an owned reference because the token owned by this
+    // struct could change while the caller is still holding the returned token.
+    fn ensure_default_service_account_oauth_token(&mut self) -> Result<String> {
+        match &self.default_service_account_oauth_token {
+            Some(token) if !token.expired() => Ok(token.token.clone()),
```

Comment:
I would write this as `=> return Ok` and terminate the `match` clause there. That lets you outdent the whole clause and makes things easier to read IMO.

Reply:
I had to make it `if let Some(token) = &self.default_service_account_oauth_token { ... }` and then put the `!token.expired()` check inside the `if` block, because the compiler says we don't yet have https://github.com/rust-lang/rust/issues/53667. But you're right, it's much neater this way.